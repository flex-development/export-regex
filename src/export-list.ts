/**
 * @file EXPORT_LIST_REGEX
 * @module export-regex/list
 */

/**
 * List `export` statement regex. Ignores matches in comments.
 *
 * @see https://regex101.com/r/KQEDdZ
 * @see https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/export#syntax
 *
 * @example
 *  import { EXPORT_LIST_REGEX } from '@flex-development/export-regex'
 *  import { dedent } from 'ts-dedent'
 *
 *  const code: string = dedent`
 *    export { defineBuildConfig, type BuildConfig }
 *    export type {
 *      JsonObject,
 *      LiteralUnion,
 *      Nullable
 *    }
 *  `
 *
 *  const print = (matches: IterableIterator<RegExpMatchArray>): void => {
 *    console.debug([...matches].map(match => omit(match, ['input'])))
 *  }
 *
 *  print(code.matchAll(EXPORT_LIST_REGEX))
 *  // [
 *  //   {
 *  //     '0': 'export { defineBuildConfig, type BuildConfig }',
 *  //     '1': undefined,
 *  //     '2': '{ defineBuildConfig, type BuildConfig }',
 *  //     index: 0,
 *  //     groups: [Object: null prototype] {
 *  //       type: undefined,
 *  //       exports: '{ defineBuildConfig, type BuildConfig }'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export type {\n  JsonObject,\n  LiteralUnion,\n  Nullable\n}',
 *  //     '1': 'type',
 *  //     '2': '{\n  JsonObject,\n  LiteralUnion,\n  Nullable\n}',
 *  //     index: 82,
 *  //     groups: [Object: null prototype] {
 *  //       type: 'type',
 *  //       exports: '{\n  JsonObject,\n  LiteralUnion,\n  Nullable\n}'
 *  //     }
 *  //   }
 *  // ]
 *
 * @const {RegExp} EXPORT_LIST_REGEX
 */
const EXPORT_LIST_REGEX: RegExp =
  /(?<=^|[\n;](?:[\t ]*(?:\w+ )?)?)export(?:(?:\s+(?<type>type)\s*)|\s*)(?<exports>{[\w\t\n\r "$'*,./{}-]+?})(?=;?[\t\n;]*?(?!(?:\n*\/\/)|(?:\n*\/\*)|(?:\s*from.*?)))/g

export default EXPORT_LIST_REGEX
