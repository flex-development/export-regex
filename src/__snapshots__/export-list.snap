// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:EXPORT_LIST_REGEX > exports > should match named export(s) in multi-line statement 1`] = `
{
  "0": "export {
  addFive,
  addFour,
  addThree,
  addTwo,
  squareFive,
  squareFour,
  squareThree,
  squareTwo
}",
  "1": undefined,
  "2": "{
  addFive,
  addFour,
  addThree,
  addTwo,
  squareFive,
  squareFour,
  squareThree,
  squareTwo
}",
  "groups": {
    "exports": "{
  addFive,
  addFour,
  addThree,
  addTwo,
  squareFive,
  squareFour,
  squareThree,
  squareTwo
}",
    "type": undefined,
  },
  "index": 0,
}
`;

exports[`unit:EXPORT_LIST_REGEX > exports > should match named export(s) in single-line statement 1`] = `
{
  "0": "export { defineBuildConfig, type BuildConfig }",
  "1": undefined,
  "2": "{ defineBuildConfig, type BuildConfig }",
  "groups": {
    "exports": "{ defineBuildConfig, type BuildConfig }",
    "type": undefined,
  },
  "index": 0,
}
`;

exports[`unit:EXPORT_LIST_REGEX > exports > should match named type export(s) in multi-line statement 1`] = `
{
  "0": "export type {
  JsonObject,
  LiteralUnion,
  Nullable
}",
  "1": "type",
  "2": "{
  JsonObject,
  LiteralUnion,
  Nullable
}",
  "groups": {
    "exports": "{
  JsonObject,
  LiteralUnion,
  Nullable
}",
    "type": "type",
  },
  "index": 0,
}
`;

exports[`unit:EXPORT_LIST_REGEX > exports > should match named type export(s) in single-line statement 1`] = `
{
  "0": "export type { default as Options }",
  "1": "type",
  "2": "{ default as Options }",
  "groups": {
    "exports": "{ default as Options }",
    "type": "type",
  },
  "index": 0,
}
`;
